### PR TITLE
Removed calls to setup things

### PR DESCRIPTION
**Pull request by aounhaider1**
Removed all calls to setup things, this should solve issue: https://github.com/eclipse/smarthome/issues/1139. However, linking and unlinking of channels is not working as expected because of the issue https://github.com/eclipse/smarthome/issues/1284.
So this should be merged only after issue 1284 is resolved.

Signed-off-by: Aoun Bukhari <bukhari@itemis.de>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
Thanks (but note that you need to rebase after I merged some of your other PRs)!
@sjka Could you work with @aounhaider1 to fix #1284, so that this PR can be merged?

**Comment by aounhaider1** [comment]
control ui is also adapted accordingly.

**Comment by kaikreuzer** [comment]
Thanks @aounhaider1!
I see a few small issues that would be nice if you could address them. See this screenshot:

bildschirmfoto 2016-04-28 um 12 06 43

 - Things without any enabled channel should not appear at all
 - The state formatting seems to be gone
 - There seems to be a separator line too much (under "Luftfeuchtigkeit" - note that the "Pressure" channel is disabled in that example)

**Comment by aounhaider1** [comment]
I've fixed the issues that you stated and have updated the pull request.

**Comment by kaikreuzer** [comment]
Removing the usage of ThingSetupManager has become a bigger refactoring.

This PR now addresses #1284, #1139, #1039.

@sjka, @maggu2810, @cdjackson: Please review and test these changes before merging. @cdjackson, I am also not in detail sure what implications it might have on HABmin. Please specifically read #1039, which explains the auto-created items. Additionally, the ThingLinkManager now automatically creates links for all non-advanced channels of newly created Things - this is meant to be the "simple" mode for OH, where users do not have to manually create items and links at all. This feature can be deactivated by setting
```
org.eclipse.smarthome.links:autoLinks=false
```
When it is deactivated, it is up to the user to create links for any channel that he wants to use ("advanced" mode, which is the normal situation for experienced OH1 users). UIs can check whether this service is active or not by accessing /rest/links/auto.

**Comment by maggu2810** [comment]
I want to give it a try, but cannot build Eclipse SmartHome after merging this branch.

```java
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.eclipse.smarthome.io.rest.core.test.item.ItemResourceOSGiTest
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.631 sec - in org.eclipse.smarthome.io.rest.core.test.item.ItemResourceOSGiTest
assert getItems with type filter works(org.eclipse.smarthome.io.rest.core.test.item.ItemResourceOSGiTest)  Time elapsed: 0.374 sec
assert getItems with tag filter works(org.eclipse.smarthome.io.rest.core.test.item.ItemResourceOSGiTest)  Time elapsed: 0.009 sec
assert addTag and removeTag works(org.eclipse.smarthome.io.rest.core.test.item.ItemResourceOSGiTest)  Time elapsed: 0.01 sec
assert expected status codes for addTag and removeTag are returned(org.eclipse.smarthome.io.rest.core.test.item.ItemResourceOSGiTest)  Time elapsed: 0.011 sec
Running org.eclipse.smarthome.io.rest.core.test.item.ItemResourceOSGiTestWithTransformTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.056 sec - in org.eclipse.smarthome.io.rest.core.test.item.ItemResourceOSGiTestWithTransformTest
assert getItems with transform bundle works(org.eclipse.smarthome.io.rest.core.test.item.ItemResourceOSGiTestWithTransformTest)  Time elapsed: 0.051 sec
Running org.eclipse.smarthome.io.rest.core.discovery.InboxResourceOSGITest
Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.086 sec <<< FAILURE! - in org.eclipse.smarthome.io.rest.core.discovery.InboxResourceOSGITest
assert that approve approves Things which are in the Inbox(org.eclipse.smarthome.io.rest.core.discovery.InboxResourceOSGITest)  Time elapsed: 0.048 sec  <<< ERROR!
groovy.lang.MissingMethodException: No signature of method: org.eclipse.smarthome.io.rest.core.discovery.InboxResource.approve() is applicable for argument types: (null, java.lang.String, java.lang.String, java.lang.Boolean) values: [null, binding:type:id, dummy_thing, false]
Possible solutions: approve(java.lang.String, java.lang.String, java.lang.String)
	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.unwrap(ScriptBytecodeAdapter.java:56)
	at org.codehaus.groovy.runtime.callsite.PojoMetaClassSite.call(PojoMetaClassSite.java:46)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:128)
	at org.eclipse.smarthome.io.rest.core.discovery.InboxResourceOSGITest.assert that approve approves Things which are in the Inbox(InboxResourceOSGITest.groovy:56)

assert that approve dont approves Things which are not in the Inbox(org.eclipse.smarthome.io.rest.core.discovery.InboxResourceOSGITest)  Time elapsed: 0.006 sec  <<< ERROR!
groovy.lang.MissingMethodException: No signature of method: org.eclipse.smarthome.io.rest.core.discovery.InboxResource.approve() is applicable for argument types: (null, java.lang.String, java.lang.String, java.lang.Boolean) values: [null, binding:type:id, dummy_thing, false]
Possible solutions: approve(java.lang.String, java.lang.String, java.lang.String)
	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.unwrap(ScriptBytecodeAdapter.java:56)
	at org.codehaus.groovy.runtime.callsite.PojoMetaClassSite.call(PojoMetaClassSite.java:46)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:128)
	at org.eclipse.smarthome.io.rest.core.discovery.InboxResourceOSGITest.assert that approve dont approves Things which are not in the Inbox(InboxResourceOSGITest.groovy:68)

Running org.eclipse.smarthome.io.rest.core.service.test.ConfigurableServiceResourceOSGiTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.075 sec - in org.eclipse.smarthome.io.rest.core.service.test.ConfigurableServiceResourceOSGiTest
assert component name fallback works(org.eclipse.smarthome.io.rest.core.service.test.ConfigurableServiceResourceOSGiTest)  Time elapsed: 0.03 sec
assert configuration management works(org.eclipse.smarthome.io.rest.core.service.test.ConfigurableServiceResourceOSGiTest)  Time elapsed: 0.032 sec
assert getConfigurableServices works(org.eclipse.smarthome.io.rest.core.service.test.ConfigurableServiceResourceOSGiTest)  Time elapsed: 0.007 sec

Results :

Tests in error: 
  InboxResourceOSGITest.assert that approve approves Things which are in the Inbox:56 » MissingMethod
  InboxResourceOSGITest.assert that approve dont approves Things which are not in the Inbox:68 » MissingMethod

Tests run: 10, Failures: 0, Errors: 2, Skipped: 0
```

File: bundles/core/org.eclipse.smarthome.core.thing/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -50,5 +50,6 @@ Export-Package: org.eclipse.smarthome.core.thing,
  org.eclipse.smarthome.core.thing.setup,
  org.eclipse.smarthome.core.thing.type,
  org.eclipse.smarthome.core.thing.util
-Service-Component: OSGI-INF/*.xml
+Service-Component: OSGI-INF/*.xml,
+ OSGI-INF/ThingLinkManager.xml
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/META-INF/MANIFEST.MF]: ThingLinkManager.xml is already catched by *.xml

File: bundles/core/org.eclipse.smarthome.core.thing/build.properties [hunk 2]
```diff
@@ -2,6 +2,7 @@ output.. = target/classes/
 bin.includes = META-INF/,\
                .,\
                OSGI-INF/,\
-               about.html
+               about.html,\
+               OSGI-INF/ThingLinkManager.xml
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/build.properties]: Should already be catched by "OSGI-INF/"

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java [hunk 3]
```diff
@@ -96,6 +101,22 @@ public void removedService(final ServiceReference reference, final Object servic
             }
         };
         thingRegistryServiceTracker.open();
+        linkRegistryServiceTracker = new ServiceTracker(this.bundleContext, ItemChannelLinkRegistry.class.getName(),
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java]: Any reason not use DS here?

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java [hunk 4]
```diff
@@ -507,7 +528,7 @@ protected Bridge getBridge() {
     protected boolean isLinked(String channelId) {
         Channel channel = thing.getChannel(channelId);
         if (channel != null) {
-            return channel.isLinked();
+            return !linkRegistry.getLinks(channel.getUID()).isEmpty();
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java]: I assume linkRegistry can be null here.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java (hunk 3 above)
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java]: Yes, the same why the thingRegistryServiceTracker isn't using DS: The whole ThingHandlers are no DS components.

Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java]: Okay, another question: Why don't the while ThingHandlers using declarative services?
It is okay to do so, just want to understand why we don't use it.

Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java]: Note that this is not related to this PR, I didn't change anything in this respect.

The reason is that handlers are created by ThingFactories - we have a handler for each thing and not just one per binding. And afair, although there are also factories for DS, this didn't really match here.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java (hunk 4 above)
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java]: fixed.

File: bundles/core/org.eclipse.smarthome.core.thing/build.properties (hunk 2 above)
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/build.properties]: fixed.

**Comment by kaikreuzer** [comment]
I incorporated @maggu2810's comments and fixed the tests.

**Comment by kaikreuzer** [comment]
Any further comments on this PR? I'd like to merge it within the next days as it otherwise makes further development on the Paper UI very difficult.

**Comment by maggu2810** [comment]
> Any further comments on this PR?

Perhaps, need to find some time to further test it...

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java (hunk 3 above)
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java]: Sorry for the question at all. It seems I missed that this is the BaseThingHandler implementation (also after you have written "ThingHandlers"). :wink: long days ATM

**Comment by maggu2810** [comment]
If my reading is correct this removes the whole localization stuff for automatic generated items.
At the moment if a german locale is used to add things, the created items' label is the localized one of the channel.
I don't like to see this function removed. Perhaps there could be a configuration of a system language that is used for the automated stuff that could be set by the UI.

**Comment by kaikreuzer** [comment]
I actually think that it is better than before. It is correct that the UI does not decide about the locale of item labels anymore - instead, the system locale is used and it is reconsidered at every startup.

Since the items are "central", it should imho not be a UI that decides what their default label is - as it should be independent of the UI used.
Also note that the label on an item is only considered to be a fallback default label and thus it imho makes sense to use the system locale. The "concrete" labels for certain UIs can be provided by ItemUIProviders.

**Comment by maggu2810** [comment]
What do you mean it is reconsidered at every startup? Am I correct that this also applies for "lost" items and my manually edit label is not changed (I am sure, but let's ask me that)?

The Paper UI already provide a page where a locale / language setting could be done.

The ESH system misses currently a mechanism to set the default language of the JVM after startup and to configure it (or I missed that code part).
Wouldn't it be a good thing to add a configuration (e.g. using config admin and so normal runtime cfg file) that is calling "Locale.setDefault(Locale)". The ESH framework needs to ensure that the locale is set in front of any other logic that is using the locale...

I think it is better to have a mechanism in the framework instead of rely on the system locale that is not easy to change on all environments and needs a restart of the framework after a change.

**Comment by kaikreuzer** [comment]
> What do you mean it is reconsidered at every startup?

I am talking about the ChannelItemProvider - it creates "virtual" items when ever a link to a non-existent item is discovered. I.e. these items are constructed upon every system start.

> Am I correct that this also applies for "lost" items and my manually edit label is not changed (I am sure, but let's ask me that)?

What are "lost" items? Manually created items are not touched; so if you create items (through DSL or ManagedItemProvider) they stay as they are, no impact there.

> The Paper UI already provide a page where a locale / language setting could be done.

Which is meant as a setting for the UI, not for the system.

> Wouldn't it be a good thing to add a configuration (e.g. using config admin and so normal runtime cfg file) that is calling "Locale.setDefault(Locale)".
> I think it is better to have a mechanism in the framework instead of rely on the system locale that is not easy to change on all environments and needs a restart of the framework after a change.

That's a "nice to have" feature for me, people use it at most once in the life of their installation, so I don't think it is really important. But I also don't have anything against introducing it. But this is imho unrelated and out of scope of this PR, I hope you agree.

**Comment by kaikreuzer** [comment]
@sjka, @maggu2810 Any further comments? As we have reached the end of this week, I really want to get this merged soon :-)

**Comment by maggu2810** [comment]
I hope we could discuss how to realize the language settings, soon.

From my side, I didn't identified any blockers

**Comment by kaikreuzer** [comment]
Thanks @maggu2810 - I take this as a lgtm and merge. If things come up from others, let's fix it afterwards.

**Comment by maggu2810** [comment]
I have waited for a comment of @sjka as you asked him for further comments, too. That is the reason I have not merged this one myself.
